beacon-chain/db/kv/state_diff_cache.go

```diff
@@ -0,0 +1,58 @@
+package kv
+
+import (
+	"math"
+	"sync"
+
+	"github.com/OffchainLabs/prysm/v6/beacon-chain/state"
+	"github.com/OffchainLabs/prysm/v6/config/params"
+	"github.com/pkg/errors"
+)
+
+const unset = math.MaxUint64
+
+type stateDiffCache struct {
+	sync.RWMutex
+	anchors []state.ReadOnlyBeaconState
+	offset  uint64
+}
+
+func newStateDiffCache() *stateDiffCache {
+	return &stateDiffCache{
+		anchors: make([]state.ReadOnlyBeaconState, len(params.StateHierarchyExponents())),
+		offset:  unset,
+	}
+}
+
+func (c *stateDiffCache) getAnchor(level int) state.ReadOnlyBeaconState {
+	c.RLock()
+	defer c.RUnlock()
+	return c.anchors[level]
+}
+
+func (c *stateDiffCache) setAnchor(level int, anchor state.ReadOnlyBeaconState) {
+	c.Lock()
+	defer c.Unlock()
+	c.anchors[level] = anchor
+}
+
+func (c *stateDiffCache) getOffset() (uint64, error) {
```

Comment:
I think this function should not return an error and we assume that unset should never be a thing when the feature is enabled. We will error out when we request the offset from db and it's not there.